IDETECT-3285  | go.mod file // indirects matching as Direct Dependenc…

jppetrakis opened the pull request:
…ie | jppetrakis

# Description

Changes to leverage information and produce a corrected go dependency map for modules.

# Github Issues

(Optional) Please link to any applicable github issues.

dterrysynopsys reviewed (approved):
Approving. I was confused about an area or two but chatted with John and believe he's adding a few more comments. It's hard to tell if this is 100% accurate as I'm no Go expert but I'm pretty convinced it is an improvement over the current code so seems like a win.

jppetrakis commented:
Fixed tests etc.

stevebillings reviewed (approved):
It may be helpful to review both the GoMod table row, and the text below the table on https://sig-confluence.internal.synopsys.com/display/integration/Detect+Audit+Log.

GoModDependencyHelper seems to combine parsing and transforming, which is something the old team had decided we wanted to avoid. Each becomes easier to understand and unit test when they are separated. Detectors vary in how well they do this. Those that haven't been touched in a while may not adhere to the pattern.

Prior to this change, the GoMod detector did this imperfectly. The transformers are in a package called "process", and none of the classes are named "*Transformer". I think that's why it got dinged in the audit. But it did separate transforming from parsing.

GoModCommandRunner.runGoModWhyOnModule() is running the command and parsing its output. It should only run the command and return the output; parsing should happen in a parser class in the parse package, simplifying unit testing.

In parsers, it's really helpful to the reader to provide an example output line in a comment.

I found myself puzzled by variable names. I have little to no idea why "rgdm" (a GoModuleDependencyHelper) is named what it is. Likewise for "m" (go mod why output line) and "rMap".

It doesn't look like the battery tests exercise "// indirects" handling, but I think it would be really valuable if it did.

jppetrakis commented:
> It may be helpful to review both the GoMod table row, and the text below the table on https://sig-confluence.internal.synopsys.com/display/integration/Detect+Audit+Log.

I did do that.  Yes it was helpful.

> 
> GoModDependencyHelper seems to combine parsing and transforming, which is something the old team had decided we wanted to avoid. Each becomes easier to understand and unit test when they are separated. Detectors vary in how well they do this. Those that haven't been touched in a while may not adhere to the pattern.
> 

My interpretation of the document's description of parsing vs. transforming is that a transform changes the objects.  In this case a List<String> (along with auxiliary data) is the input and the output is a List<String>, with only the data itself changed.  So from what the document says, it's not really transforming.


> Prior to this change, the GoMod detector did this imperfectly. The transformers are in a package called "process", and  none of the classes are named "*Transformer". I think that's why it got dinged in the audit. But it did separate transforming from parsing.
> 
> GoModCommandRunner.runGoModWhyOnModule() is running the command and parsing its output. It should only run the command and return the output; parsing should happen in a parser class in the parse package, simplifying unit testing.
> 
> In parsers, it's really helpful to the reader to provide an example output line in a comment.
> 

runGoModWhyOnModule does happen to be a transformer, since it takes a List<String> and returns a HashMap<String, List<String>>.  It does not actually change the data but it HAS to read the data in order to go from a list to a map... thus I moved that aspect of it out to a new class, and now use runGoModWhy

Since I need to do a go mod why on each module to get its lineage, and since the command writes an output line every time it's called... it seemed better to call the why command once and alter the list.

> I found myself puzzled by variable names. I have little to no idea why "rgdm" (a GoModuleDependencyHelper) is named what it is. Likewise for "m" (go mod why output line) and "rMap".

Sorry, that's because I renamed the class and didn't update the variable names.

> 
> It doesn't look like the battery tests exercise "// indirects" handling, but I think it would be really valuable if it did.

jppetrakis commented:
merging since approved.
